### DISCUSSION TITLE
v5 custom repository circular dependency

### DISCUSSION THREAD

**Discussion by alexandre-abrioux** [mikro-orm/mikro-orm#4089] [Q&A] [answered]
Hi! First of all, thank you for maintaining this fantastic project 🙂

It looks like the new way of defining custom repositories in v5 introduces circular dependencies.

```ts
// simplified content of entity.ts
import { UserRepository } from "./repository";

@Entity({ customRepository: () => UserRepository })
export class User {}
```

```ts
// simplified content of repository.ts
import { User } from "./entity";

export class UserRepository extends EntityRepository<User> {}
```

As you can see, both modules depend on each other.

One solution could be to declare both classes in the same file, but knowing that the content of those classes is often sizeable, it is, at least in my opinion, preferable to store them in different modules.

The old way of defining custom repositories (with the `@Repository()` decorator) didn't have that flaw.

Let me know if you need more details. Cheers!

**Versions**

| Dependency | Version |
| - | - |
| mikro-orm | v5 |

**Comment by B4nan** [comment, marked as answer]
The old way was much more flawed, as it didn't work unless you actually required your repository. And given most of the people want to use `EntityRepositoryType` symbol they ended up with the same cycle.

Do you have an actual problem with this approach? It works fine in all the example repos and tests. This kind of cycles are common in entity definitions in general, every bidirectional relation will have it.

**Reply by alexandre-abrioux** [reply]
Hi @B4nan, thanks for the answer! We have no problem per se: we're just detecting circular dependencies in our CI and getting warnings, so I wanted to let you know. But that's OK, like you said, circular references are standard when dealing with entity relations, so it shouldn't be an issue. I'm closing this. Have a good day! 🙂
